llvm/lib/Transforms/Yk/SplitBlocksAfterCalls.cpp

```diff
@@ -0,0 +1,115 @@
+//===- SplitBlocksAfterCalls.cpp -===//
+//
+// Makes function calls effectively terminators by splitting blocks after each
+// call. This ensures that there can only be at most one call per block. This
+// is used in order to detect recursion and external function calls within a
+// trace.
+
+#include "llvm/Transforms/Yk/SplitBlocksAfterCalls.h"
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/Verifier.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Pass.h"
+#include "llvm/Transforms/Yk/LivenessAnalysis.h"
+
+#include <map>
+
+#define DEBUG_TYPE "yk-splitblocksaftercalls"
+
+using namespace llvm;
+
+namespace llvm {
+void initializeYkSplitBlocksAfterCallsPass(PassRegistry &);
+} // namespace llvm
+
+namespace {
+
+class YkSplitBlocksAfterCalls : public ModulePass {
+public:
+  static char ID;
+  YkSplitBlocksAfterCalls() : ModulePass(ID) {
+    initializeYkSplitBlocksAfterCallsPass(*PassRegistry::getPassRegistry());
+  }
+
+  bool runOnModule(Module &M) override {
+    LLVMContext &Context = M.getContext();
+
+    for (Function &F : M) {
+      if (F.empty()) // skip declarations.
+        continue;
+      std::vector<BasicBlock *> Todo;
```

Comment:
Interesting you used a work list.

Is there any benefit to just looping over all blocks?

Reply:
Good answer! Can we put a not in a comment to remind our future selves?

Reply:
[3516bad](https://github.com/ykjit/ykllvm/pull/108/commits/3516badd72adc3f946189d53293e11e8442c9347)